src/backend/native_gpu_future.rs

```diff
@@ -24,43 +28,58 @@ impl<T> Future for GpuFuture<T>
     type Output = T;
 
     fn poll(self: Pin<&mut Self>, context: &mut Context) -> Poll<Self::Output> {
-        // grab a clone of the Arc
-        let arc = Arc::clone(&self.get_mut().inner);
+        let inner = &self.get_mut().inner;
 
         // grab the device id and set the waker, but release the lock, so that the native callback can write to it
         let device_id = {
-            let mut inner = arc.lock().unwrap();
-            inner.waker.replace(context.waker().clone());
+            let mut inner = inner.lock().unwrap();
+
+            if let Some(WakerOrResult::Result(res)) = inner.waker_or_result.take() {
+                return Poll::Ready(res);
+            }
+
+            inner.waker_or_result.replace(WakerOrResult::Waker(context.waker().clone()));
+
             inner.id
         };
 
         // polling the device should trigger the callback
-        wgn::wgpu_device_poll(device_id, true);
+        wgn::wgpu_device_poll(device_id, false);
 
         // now take the lock again, and check whether the future is complete
-        let mut inner = arc.lock().unwrap();
-        match inner.result.take() {
-            Some(value) => Poll::Ready(value),
-            _ => Poll::Pending,
+        let mut inner = inner.lock().unwrap();
+
+        match inner.waker_or_result.take() {
+            Some(WakerOrResult::Result(res)) => Poll::Ready(res),
+            value @ _ => {
```

Comment:
that's a strange pattern, could just be `value => { ... }`

Reply:
Oops, didn't realize I could just do that.